lib/mongo/server/connection_pool/available_stack.rb

```diff
@@ -119,121 +120,120 @@ def initialize(options = {}, &block)
-          check_count_invariants
+          check_count_invariants(false)
           mutex.synchronize do
-            queue.each{ |connection| connection.disconnect! }
-            @pool_size -= queue.length
-            if @pool_size < 0
-              # This should never happen
-              log_warn("ConnectionPool::Queue: connection accounting problem")
-              @pool_size = 0
+            connections.each do |connection|
+              connection_removed
+              connection.disconnect!
+
+              publish_cmap_event(
+                  Monitoring::Event::ConnectionClosed.new(
+                      Monitoring::Event::ConnectionClosed::POOL_CLOSED,
+                      @address,
+                      connection.id,
+                  ),
+              )
             end
-            queue.clear
+
+            connections.clear
             @generation += 1
             while @pool_size < min_size
               @pool_size += 1
-              queue.unshift(@block.call(@generation))
+              connections.unshift(@block.call(@generation))
             end
             true
           end
         ensure
           check_count_invariants
         end
 
-        # Enqueue a connection in the queue.
+        # Push a connection onto the stack.
         #
-        # Only connections created by this queue should be enqueued
-        # back into it, however the queue does not verify whether it
-        # originally created the connection being enqueued.
+        # Only connections created by this stack should be pushed
+        # back into it, however the stack does not verify whether it
+        # originally created the connection being pushed.
         #
-        # If linting is enabled (see Mongo::Lint), attempting to enqueue
+        # If linting is enabled (see Mongo::Lint), attempting to push
         # connections beyond the pool's capacity will raise Mongo::Error::LintError
         # (since some of those connections must not have originated from
-        # the queue into which they are being enqueued). If linting is
-        # not enabled, the queue can grow beyond its max size with undefined
+        # the stack into which they are being pushed). If linting is
+        # not enabled, the stack can grow beyond its max size with undefined
         # results.
         #
-        # @example Enqueue a connection.
-        #   queue.enqueue(connection)
+        # @example Push a connection.
+        #   stack.push(connection)
         #
         # @param [ Mongo::Server::Connection ] connection The connection.
         #
         # @since 2.0.0
-        def enqueue(connection)
-          check_count_invariants
+        def push(connection)
+          check_count_invariants(false)
           mutex.synchronize do
             if connection.generation == @generation
-              queue.unshift(connection.record_checkin!)
+              connections.unshift(connection.record_checkin!)
               resource.broadcast
```

Comment:
This broadcast only happens if the connection is of the same generation. If the connection is stale, there is no broadcast (for the thread waiting in get_connection) but the stack size may have reduced below the max # of allowed connections, therefore the waiting thread should check that and potentially create a new connection immediately.

Reply:
Fixed